### DISCUSSION TITLE
Mutate data outside from react component/hook

### DISCUSSION THREAD

**Discussion by NWPoul** [TanStack/query#6743] [General] [answered]
How we can mutate data outside of react component?
Searching offer not working solution to use 'mutate' method of application's instance of QueryClient...
which looks pretty logical but instance of QueryClient does not contain such a method

**Comment by TkDodo** [comment]
you mean fire off a `mutation` ? There isn't much gain in doing that outside of react. You can't take advantage of the lifecycle management, so you can just as easily fire off a request with just fetch / axios.

But it's definitely possible:

https://github.com/TanStack/query/blob/5d29f8ff47958a2349fdc4ca682c3630beeabc9d/packages/query-core/src/tests/utils.ts#L35-L44

**Comment by adi-doorloop** [comment]
@TkDodo How do you do it for V4? `mutation.execute()` doesn't receive variables, or maybe just a type error?

**Reply by denisborovikov** [reply]
iirc, `build` accepts variables in the 3rd argument `.build(queryClient, options, { variables })`

**Comment by RBilly** [comment]
Hi, I've got quite the edge case where I also need to fire a mutation outside of react.
So I've a websocket handler outside of react and when some type of message is received, I want to fire an existing mutation.
So I have tried this in the function called by the WS handler:
```js
export const updateProcess = async () => {
  queryClient.getMutationCache().find({ mutationKey })?.execute();
};
```
and
```js
export const updateProcess = async () => {
  const mutation = queryClient.getMutationCache().find({ mutationKey });
  if (mutation ) {
    queryClient.getMutationCache().build(
         queryClient, mutation.options, mutation.state.variables
    ).execute();
  }
};
```
but none of them works. In both cases the onMutate function throw an error saying that the expected params is undefined.
but when I display `mutation.state.variables`, it is correctly defined.

Is there any way to do this or is it a stupid idea ?

Btw I'm using react-query v4

**Reply by RBilly** [reply]
Since it was really bothering me, I went looking into the code and found the answer as to why it does not work.
When displaying the content of the mutation retrieved from the cache, the state contains the expected variables.
But, in the `execute` function from the mutation class file there this line of code (l. 202):
`this.dispatch({ type: 'loading', variables: this.options.variables! })`
As you can see, it takes variables from the mutation options and that's where things go wrong.
Because the dispatch will set the mutation state variables with options variables passed as action.variables parameter:
```js
case 'loading':
          return {
            ...state,
            context: action.context,
            data: undefined,
            failureCount: 0,
            failureReason: null,
            error: null,
            isPaused: !canFetch(this.options.networkMode),
            status: 'loading',
            variables: action.variables, // <--- here
          }
```
But after the mutation is fired, the "variables" field of options does not exists anymore.
And when onMutate is called with mutation state variables after the loading dispatch, it's actually undefined now...
So it's impossible to fire it again just by calling execute. First we need to set the options.variables using the state.variables.
```js
export const updateProcess = async () => {
  const mutation = queryClient.getMutationCache().find({ mutationKey });
  if (mutation ) {
    mutation.setOptions({...mutation.options, variables: {...mutation.state.variables}})
    await mutation.execute();
  }
};
```

Now I'm able to work around the issue, but my question is why the options.variables is removed. Is it meant to be like this for a reason ?

**Comment by romellem** [comment]
I want to rebump this for query@v5. I too have the need to imperatively add data to the mutationCache (we have to react to a redux event in a [saga](https://redux-saga.js.org/) (I know, I know)). For QueryCache, this is handled by `queryClient.fetchQuery` + `queryOptions()`. However I'm not sure how to do the same for mutations.

I was a bit surprised to see the docs for `MutationCache` not include the `build` + `execute`, or `find` methods.

Should I take these methods being undocumented to mean these are meant to be private? How "dangerous" is it to try and call these manually?

https://github.com/TanStack/query/blob/f6ae1ccdf74a61ea15e845d4643e6d4b0a0b7bdf/packages/query-core/src/mutationCache.ts#L105

https://github.com/TanStack/query/blob/f6ae1ccdf74a61ea15e845d4643e6d4b0a0b7bdf/packages/query-core/src/mutation.ts#L176

**Reply by TkDodo** [reply]
this is still the best way:

https://github.com/TanStack/query/discussions/6743#discussioncomment-8265716

they are part of the public, low-level `query-core` API. Most of it isn’t documented but it’s also not going away. Feel free to update the docs if you want.

**Reply by romellem** [reply]
Thanks for the confirmation! And I agree, in an ideal world, most users wouldn't need to use the low-level APIs. My world is definitely less than ideal 😄

**Comment by amasen02** [comment]
In TanStack Query v5, the official, public primitive for executing and observing mutations outside of React (such as inside a Redux saga, WebSocket listener, or background service) is **`MutationObserver`** from `@tanstack/query-core`.

Just as `useQuery` is a thin React wrapper around `QueryObserver`, `useMutation` is simply a wrapper around `MutationObserver` (see `packages/react-query/src/useMutation.ts`). 

### Why avoid `MutationCache.build()` and `.execute()` manually?
- **`MutationCache.build()`** is an internal cache method. Calling `.execute()` directly bypasses default configuration inheritance (`queryClient.getMutationDefaults()`), retry orchestration, and observer lifecycle management.
- **`MutationCache.find()`** only searches for mutations that are already created or resumed in the cache; it is not an execution entrypoint.

---

### The Recommended Solution: `MutationObserver`

You can instantiate a `MutationObserver` directly with your `QueryClient` instance:

```ts
import { MutationObserver } from '@tanstack/query-core'

// 1. Instantiate observer with your existing queryClient
const observer = new MutationObserver(queryClient, {
  mutationKey: ['updateTodo'],
  mutationFn: async (variables: { id: string; title: string }) => {
    const res = await fetch(`/api/todos/${variables.id}`, {
      method: 'PATCH',
      body: JSON.stringify(variables),
    })
    if (!res.ok) throw new Error('Network error')
    return res.json()
  },
  onSuccess: (data) => {
    // Standard cache invalidations work seamlessly
    queryClient.invalidateQueries({ queryKey: ['todos'] })
  },
})

// 2. Execute imperatively:
// Like mutateAsync, observer.mutate() returns a promise:
try {
  const result = await observer.mutate({ id: '123', title: 'New title' })
  console.log('Success:', result)
} catch (error) {
  console.error('Failed:', error)
}
```

---

### Usage in Redux Sagas / External Event Handlers

In a Redux saga or external handler, you can call `observer.mutate`:

```ts
import { call } from 'redux-saga/effects'
import { MutationObserver } from '@tanstack/query-core'
import { queryClient } from './queryClient'
import { updateTodoOptions } from './mutations'

function* onSocketEventSaga(action) {
  const observer = new MutationObserver(queryClient, updateTodoOptions)
  try {
    const data = yield call([observer, observer.mutate], action.payload)
    console.log('Saga mutation complete:', data)
  } catch (error) {
    console.error('Saga mutation error:', error)
  }
}
```

---

### Advantages over low-level cache manipulation:
1. **Global Cache & UI Synchronization**: Because `MutationObserver` registers the mutation with the global `mutationCache`, any React components using `useMutationState({ filters: { mutationKey: ['updateTodo'] } })` will automatically observe the active `isPending` status, loading spinners, and result states in real-time.
2. **Retry & Backoff**: Automatically inherits any `retry` and `retryDelay` policies defined on the mutation options or client defaults.
3. **Public API Stability**: `MutationObserver` is a documented, stable export of `@tanstack/query-core` across v4 and v5.
